decoders/netflow/templates/file/file.go

```diff
@@ -0,0 +1,199 @@
+		Type string
+		Data interface{} `json:"-"`
+	}
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+
+	switch s.Type {
+	case "NFv9OptionsTemplateRecord":
+		newS := new(struct{
+			Type string
+			Data netflow.NFv9OptionsTemplateRecord
+		})
+		if err := json.Unmarshal(b, newS); err != nil {
+			return err
+		}
+		d.Type = newS.Type
+		d.Data = newS.Data
+	case "TemplateRecord":
+		newS := new(struct{
+			Type string
+			Data netflow.TemplateRecord
+		})
+		if err := json.Unmarshal(b, newS); err != nil {
+			return err
+		}
+		d.Type = newS.Type
+		d.Data = newS.Data
+	case "IPFIXOptionsTemplateRecord":
+		newS := new(struct{
+			Type string
+			Data netflow.IPFIXOptionsTemplateRecord
+		})
+		if err := json.Unmarshal(b, newS); err != nil {
+			return err
+		}
+		d.Type = newS.Type
+		d.Data = newS.Data
+	}
+
+	return nil
+}
+
+type TemplateFile struct {
+	Templates []*TemplateFileObject `json:"templates"`
+}
+
+func (f *TemplateFile) Add(key *templates.TemplateKey, data interface{}) {
+	var typeName string
+
+	switch data.(type) {
+	case netflow.NFv9OptionsTemplateRecord:
+		typeName = "NFv9OptionsTemplateRecord"
+	case netflow.TemplateRecord:
+		typeName = "TemplateRecord"
+	case netflow.IPFIXOptionsTemplateRecord:
+		typeName = "IPFIXOptionsTemplateRecord"
+	default:
+		return
+	}
+
+	f.Templates = append(f.Templates, &TemplateFileObject{
+		Key:  key,
+		Data: &TemplateFileData{
+			Type: typeName,
+			Data:data,
+		},
+	})
+}
+
+func NewTemplateFile() *TemplateFile {
+	return &TemplateFile{
+		Templates: make([]*TemplateFileObject, 0),
+	}
+}
+
+type FileDriver struct {
+	memDriver *memory.MemoryDriver
+	path string
+}
+
+func (d *FileDriver) Prepare() error {
+	d.memDriver = memory.Driver
+	flag.StringVar(&d.path, "netflow.templates.file.path", "./templates.json", "Path of file to store templates")
+	return nil
+}
+
+func (d *FileDriver) Init(ctx context.Context) error {
+	var err error
+	if err = d.memDriver.Init(ctx); err != nil {
+		return err
+	}
+
+	f, err := os.OpenFile(d.path, os.O_RDWR|os.O_CREATE, 0755)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	dec := json.NewDecoder(f)
+	tf := NewTemplateFile()
+	if err = dec.Decode(tf); err != nil {
+		// log error
+	}
+	for _, template := range tf.Templates {
+		if err := d.memDriver.AddTemplate(ctx, template.Key, template.Data.Data); err != nil {
+			// log error
+			continue
+		}
+	}
+
+	return nil
+}
+
+func (d *FileDriver) Close(ctx context.Context) error {
+	if err := d.memDriver.Close(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *FileDriver) ListTemplates(ctx context.Context, ch chan *templates.TemplateKey) error {
+	return d.memDriver.ListTemplates(ctx, ch)
+}
+
+func (d *FileDriver) AddTemplate(ctx context.Context, key *templates.TemplateKey, template interface{}) error {
+	if err := d.memDriver.AddTemplate(ctx, key, template); err != nil {
+		return err
+	}
+
+	tf := NewTemplateFile()
+
+	ch := make(chan *templates.TemplateKey, 5)
```

Comment:
There could be a problem with >1 worker.
When there is >1 worker and `len(templates) > len(ch)`, The template update could result in a potential deadlock. (There could be other cases too)

Here's a stacktrace since the issue:
https://gist.github.com/shyam334/970eece9ff4754d14fad69a4168adb58

Reply:
Haven't given much thought, But is it worthwhile making the template-updates-channel a common channel across the  workers.
(We could write from a single place and could enable controlling how-often we flush updates to disk etc)

Reply:
Oh good catch. Will have a look. Thank you